### ISSUE TITLE
Default mapper with SortedSet throws ClassCastException

### ISSUE THREAD

**Issue by yonatang** [orika-mapper/orika#58]
(Gist test case: https://gist.github.com/f537ae568a26d08163b8)

Using Orika 1.4.5 to map a bean with SortedSet, it throws the following exception:
```
ma.glasnost.orika.MappingException: While attempting the following mapping:
sourceClass = class orika.SortedSetTest$Foo
destinationType = orika.SortedSetTest.Bar
resolvedStrategy = InstantiateAndUseCustomMapperStrategy<Foo, Bar> {customMapper: GeneratedMapper<Foo, Bar> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [String] }, unenhancer: ma.glasnost.orika.unenhance.BaseUnenhancer@56528192, objectFactory: DefaultConstructorObjectFactory<Bar>}
Error occurred: java.lang.ClassCastException: java.util.LinkedHashSet cannot be cast to java.util.SortedSet
-----begin dump of current state-----------------------------
Registered object factories: 1 (approximate size: 27.3 kB)
  [Bar] : {Foo=DefaultConstructorObjectFactory<Bar>}
-------------------------------------------------------------
Registered mappers: 1 (approximate size: 887.1 kB)
  [0] : GeneratedMapper<Foo, Bar> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [String] }
-------------------------------------------------------------
Registered concrete types: 5 (approximate size: 136.0 kB)
  [interface java.util.Collection] : ArrayList<Object>
  [interface java.util.List] : ArrayList<Object>
  [interface java.util.Map$Entry] : MapEntry<Object, Object>
  [interface java.util.Set] : LinkedHashSet<Object>
  [interface java.util.Map] : LinkedHashMap<Object, Object>
-------------------------------------------------------------
Resolved strategies: 1 (approximate size: 1,012.4 kB)
{source: Foo, dest: Bar, in-place:false}: InstantiateAndUseCustomMapperStrategy<Foo, Bar> {customMapper: GeneratedMapper<Foo, Bar> {usedConverters: [], usedMappers: [], usedMapperFacades: [], usedTypes: [String] }, unenhancer: ma.glasnost.orika.unenhance.BaseUnenhancer@56528192, objectFactory: DefaultConstructorObjectFactory<Bar>}
-------------------------------------------------------------
Unenhance strategy: ma.glasnost.orika.unenhance.BaseUnenhancer@56528192
-----end dump of current state-------------------------------
	at ma.glasnost.orika.impl.ExceptionUtility.newMappingException(ExceptionUtility.java:55)
	at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:745)
	at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:714)
	at orika.SortedSetTest.shouldMapSortedSet(SortedSetTest.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:78)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:212)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:68)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
Caused by: java.lang.ClassCastException: java.util.LinkedHashSet cannot be cast to java.util.SortedSet
	at ma.glasnost.orika.generated.Orika_Bar_Foo_Mapper44633627972572$0.mapAtoB(Orika_Bar_Foo_Mapper44633627972572$0.java)
	at ma.glasnost.orika.impl.mapping.strategy.UseCustomMapperStrategy.map(UseCustomMapperStrategy.java:67)
	at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:735)
	... 29 more
```

It seems like SortedSet is not identified by the default mapper, and it tries to use LinkedHashSet instead.

**Comment by elaatifi** [comment]
Thanks for reporting this issue.

This case should be fixed by registering `TreeSet` as default implementation for `SortedSet`:
```java
factory.registerConcreteType(SortedSet.class, TreeSet.class);
```
Unfortunately the problem lay in `MultiOccurrenceVariableRef#newInstance` [see here]  (https://github.com/orika-mapper/orika/blob/master/core/src/main/java/ma/glasnost/orika/impl/generator/MultiOccurrenceVariableRef.java#L263) which is kind of dirty method for creating new instances of collection without taking care of registered concrete types for a given interface. This have to be fixed :)

As a matter of workaround you can register a custom mapper for `SortedSet`:

```java
factory.registerMapper(new CustomMapper<SortedSet<Object>, SortedSet<Object>>() {
    @Override
    public void mapAtoB(SortedSet<Object> a, SortedSet<Object> b, MappingContext context) {
        b.clear();
        b.addAll(mapper.mapAsList(a, context.getResolvedDestinationType().getNestedType(1).getRawType(), context));
    }

    @Override
    public void mapBtoA(SortedSet<Object> a, SortedSet<Object> b, MappingContext context) {
        a.clear();
        b.addAll(mapper.mapAsList(b, context.getResolvedDestinationType().getNestedType(1).getRawType(), context));
    }
});
```

**Comment by yonatang** [comment]
Thanks for the workaround! I did struggled around with the ```registerConcreteType``` unsuccessfully, trying to get it working. Just for the benefit of the community, you have a few typos in your snippet. The correct code is:
```
factory.registerConcreteType(SortedSet.class, TreeSet.class);
factory.registerMapper(new CustomMapper<SortedSet<Object>, SortedSet<Object>>() {
    @Override
    public void mapAtoB(SortedSet<Object> a, SortedSet<Object> b, MappingContext context) {
        b.clear();
        b.addAll(mapper.mapAsList(a, context.getResolvedDestinationType().getNestedType(0).getRawType(), context));
    }

    @Override
    public void mapBtoA(SortedSet<Object> a, SortedSet<Object> b, MappingContext context) {
        a.clear();
        a.addAll(mapper.mapAsList(b, context.getResolvedDestinationType().getNestedType(0).getRawType(), context));
    }
});
```

**Comment by elaatifi** [comment]
Awesome !

**Comment by elaatifi** [comment]
*UPDATE* : Should be fixed in **1.4.6**
Adding support of SortedSet and SortedMap interfaces.
